net/net.go

```diff
@@ -70,15 +70,23 @@ func AddIPOffset(base *big.Int, offset int) net.IP {
 // RangeSize returns the size of a range in valid addresses.
 // returns the size of the subnet (or math.MaxInt64 if the range size would overflow int64)
 func RangeSize(subnet *net.IPNet) int64 {
```

Comment:
I wonder if we can just get rid of the need for this function?  We need to keep it as an API, but it's only used in a handful of places in k/k - could we kill those?

Reply:
that will be ideal

Reply:
I looked quickly and it spins out a fair bit - all the allocators need to take `hostBits, reservedIPs int` instead of `max int`.  Once you get into the bitmap allocator you can turn that into a real number.

We detect IPv6 and limit it to 16 bits, but we don't limit v4 (as far as I can tell) which means there COULD BE a cluster out ther using a /8 as service CIDR.  Gross.

Then it ripples into the Used() and Free() methods.  Might be easier to just use a bignum or a uint128 library?

Reply:
Partial example:

```diff
diff --git a/pkg/registry/core/service/allocator/bitmap.go b/pkg/registry/core/service/allocator/bitmap.go
index c176c578cdc..d4508ed6f6d 100644
--- a/pkg/registry/core/service/allocator/bitmap.go
+++ b/pkg/registry/core/service/allocator/bitmap.go
@@ -60,8 +60,8 @@ type bitAllocator interface {
 }
 
 // NewAllocationMap creates an allocation bitmap using the random scan strategy.
-func NewAllocationMap(max int, rangeSpec string) *AllocationBitmap {
-	return NewAllocationMapWithOffset(max, rangeSpec, 0)
+func NewAllocationMap(hostBits, reserved int, rangeSpec string) *AllocationBitmap {
+	return NewAllocationMapWithOffset(hostBits, reserved, rangeSpec, 0)
 }
 
 // NewAllocationMapWithOffset creates an allocation bitmap using a random scan strategy that
@@ -69,7 +69,7 @@ func NewAllocationMap(max int, rangeSpec string) *AllocationBitmap {
 // The first block of values will not be used for random value assigned by the AllocateNext()
 // method until the second block of values has been exhausted.
 // The offset value must be always smaller than the bitmap size.
-func NewAllocationMapWithOffset(max int, rangeSpec string, offset int) *AllocationBitmap {
+func NewAllocationMapWithOffset(hostBits, reserved int, rangeSpec string, offset int) *AllocationBitmap {
 	a := AllocationBitmap{
 		strategy: randomScanStrategyWithOffset{
 			rand:   rand.New(rand.NewSource(time.Now().UnixNano())),
@@ -77,7 +77,7 @@ func NewAllocationMapWithOffset(max int, rangeSpec string, offset int) *Allocati
 		},
 		allocated: big.NewInt(0),
 		count:     0,
-		max:       max,
+		max:       (1 << hostBits) - reserved,
 		rangeSpec: rangeSpec,
 	}
 
diff --git a/pkg/registry/core/service/allocator/interfaces.go b/pkg/registry/core/service/allocator/interfaces.go
index 41f966858b7..733fe8c7bad 100644
--- a/pkg/registry/core/service/allocator/interfaces.go
+++ b/pkg/registry/core/service/allocator/interfaces.go
@@ -46,4 +46,4 @@ type Snapshottable interface {
 
 type AllocatorFactory func(max int, rangeSpec string) (Interface, error)
 
-type AllocatorWithOffsetFactory func(max int, rangeSpec string, offset int) (Interface, error)
+type AllocatorWithOffsetFactory func(hostBits, reserved int, rangeSpec string, offset int) (Interface, error)
diff --git a/pkg/registry/core/service/ipallocator/bitmap.go b/pkg/registry/core/service/ipallocator/bitmap.go
index 02878ea48c4..49847f01b10 100644
--- a/pkg/registry/core/service/ipallocator/bitmap.go
+++ b/pkg/registry/core/service/ipallocator/bitmap.go
@@ -46,8 +46,11 @@ type Range struct {
 	net *net.IPNet
 	// base is a cached version of the start IP in the CIDR range as a *big.Int
 	base *big.Int
-	// max is the maximum size of the usable addresses in the range
-	max int
+	// hostBits is the number of address bits which indicate hosts, as opposed
+	// to the CIDR prefix.
+	hostBits int
+	// reserved is the number or IPs that are not usable in this range
+	reserved int
 	// family is the IP family of this range
 	family api.IPFamily
 
@@ -58,48 +61,55 @@ type Range struct {
 
 var _ Interface = (*Range)(nil)
 
+func rangeSizeBits(cidr *net.IPNet) int {
+	ones, bits := cidr.Mask.Size()
+	return bits - ones
+}
+
 // New creates a Range over a net.IPNet, calling allocatorFactory to construct the backing store.
 func New(cidr *net.IPNet, allocatorFactory allocator.AllocatorWithOffsetFactory) (*Range, error) {
-	max := netutils.RangeSize(cidr)
+	hostBits := rangeSizeBits(cidr)
 	base := netutils.BigForIP(cidr.IP)
 	rangeSpec := cidr.String()
 	var family api.IPFamily
 
+	reserved := 0
 	if netutils.IsIPv6CIDR(cidr) {
 		family = api.IPv6Protocol
 		// Limit the max size, since the allocator keeps a bitmap of that size.
-		if max > 65536 {
-			max = 65536
+		if hostBits > 16 {
+			hostBits = 16
 		}
 	} else {
 		family = api.IPv4Protocol
 		// Don't use the IPv4 network's broadcast address, but don't just
 		// Allocate() it - we don't ever want to be able to release it.
-		max--
+		reserved++
 	}
 
 	// Don't use the network's ".0" address, but don't just Allocate() it - we
 	// don't ever want to be able to release it.
 	base.Add(base, big.NewInt(1))
-	max--
+	reserved++
 
 	// cidr with whole mask can be negative
-	if max < 0 {
-		max = 0
+	if hostBits == 0 {
+		reserved = 0
 	}
 
 	r := Range{
-		net:     cidr,
-		base:    base,
-		max:     maximum(0, int(max)),
-		family:  family,
-		metrics: &emptyMetricsRecorder{}, // disabled by default
+		net:      cidr,
+		base:     base,
+		hostBits: hostBits,
+		reserved: reserved,
+		family:   family,
+		metrics:  &emptyMetricsRecorder{}, // disabled by default
 	}
 
 	offset := calculateRangeOffset(cidr)
 
 	var err error
-	r.alloc, err = allocatorFactory(r.max, rangeSpec, offset)
+	r.alloc, err = allocatorFactory(r.hostBits, r.reserved, rangeSpec, offset)
 	if err != nil {
 		return nil, err
 	}
@@ -108,8 +118,8 @@ func New(cidr *net.IPNet, allocatorFactory allocator.AllocatorWithOffsetFactory)
 
 // NewInMemory creates an in-memory allocator.
 func NewInMemory(cidr *net.IPNet) (*Range, error) {
-	return New(cidr, func(max int, rangeSpec string, offset int) (allocator.Interface, error) {
-		return allocator.NewAllocationMapWithOffset(max, rangeSpec, offset), nil
+	return New(cidr, func(hostBits, reserved int, rangeSpec string, offset int) (allocator.Interface, error) {
+		return allocator.NewAllocationMapWithOffset(hostBits, reserved, rangeSpec, offset), nil
 	})
 }
 
@@ -129,13 +139,6 @@ func NewFromSnapshot(snap *api.RangeAllocation) (*Range, error) {
 	return r, nil
 }
 
-func maximum(a, b int) int {
-	if a > b {
-		return a
-	}
-	return b
-}
-
 // Free returns the count of IP addresses left in the range.
 func (r *Range) Free() int {
 	return r.alloc.Free()
@@ -356,26 +359,26 @@ func calculateIPOffset(base *big.Int, ip net.IP) int {
 func calculateRangeOffset(cidr *net.IPNet) int {
 	// default values for min(max($min, cidrSize/$step), $max)
 	const (
-		min  = 16
-		max  = 256
-		step = 16
+		min  = 4
+		max  = 8
+		step = 4
 	)
 
-	cidrSize := netutils.RangeSize(cidr)
+	hostBits := rangeSizeBits(cidr)
 	// available addresses are always less than the cidr size
 	// A /28 CIDR returns 16 addresses, but 2 of them, the network
 	// and broadcast addresses are not available.
-	if cidrSize <= min {
+	if hostBits <= min {
 		return 0
 	}
+	if hostBits > 8 {
+		return 1 << max
+	}
 
-	offset := cidrSize / step
+	offset := hostBits - step
 	if offset < min {
 		return min
 	}
-	if offset > max {
-		return max
-	}
 	return int(offset)
 }
 
```

Reply:
> We detect IPv6 and limit it to 16 bits, but we don't limit v4 (as far as I can tell)

It's limited to 20 bits by [kube-apiserver command-line arg validation](https://github.com/kubernetes/kubernetes/blob/v1.26.2/cmd/kube-apiserver/app/options/validation.go#L38)

Reply:
yeah, ipv4 is until /12 IIRC